src/libopensc/card-laser.c

```diff
@@ -0,0 +1,2697 @@
+
+/* our operations table with overrides */
+static struct sc_card_operations laser_ops;
+
+static struct sc_card_driver laser_drv = {
+		"JaCarta PKI driver",
+		"laser",
+		&laser_ops,
+		NULL, 0, NULL};
+
+static struct sc_atr_table laser_known_atrs[] = {
+		{"FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF",
+			"3B:DC:18:FF:81:91:FE:1F:C3:80:73:C8:21:13:66:01:06:11:59:00:01:28",
+			"JaCarta PKI",								   SC_CARD_TYPE_ALADDIN_LASER, 0, NULL},
+
+		{"FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF",
+			"3B:DC:18:FF:81:91:FE:1F:C3:80:73:C8:21:13:66:01:0B:03:52:00:05:38",
+			"JaCarta PKI",								   SC_CARD_TYPE_ALADDIN_LASER, 0, NULL},
+
+		{"FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF",
+			"3B:8C:80:01:80:73:C8:21:13:66:01:06:11:59:00:01:2C",
+			"JaCarta PKI/BIO",							       SC_CARD_TYPE_ALADDIN_LASER, 0, NULL},
+
+		{"FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF",
+			"3B:6C:00:FF:80:73:C8:21:13:66:01:06:11:59:00:01",
+			"JaCarta PKI/BIO",							       SC_CARD_TYPE_ALADDIN_LASER, 0, NULL},
+
+		{"FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF:FF",
+			"3B:9F:11:81:11:3D:00:11:00:00:00:00:00:00:00:00:00:00:00:00:00:32",
+			"JaCarta-2 PKI",								 SC_CARD_TYPE_ALADDIN_LASER, 0, NULL},
+
+		{NULL,								NULL, NULL, 0,			      0, NULL}
+};
+
+static struct sc_aid laser_aid = {
+		{0xA0, 0x00, 0x00, 0x01, 0x64, 0x4C, 0x41, 0x53, 0x45, 0x52, 0x00, 0x01},
+		12
+};
+
+unsigned char laser_ops_df[6] = {
+		SC_AC_OP_CREATE, SC_AC_OP_CREATE_DF, SC_AC_OP_ADMIN, SC_AC_OP_DELETE_SELF, SC_AC_OP_ACTIVATE, SC_AC_OP_DEACTIVATE};
+unsigned char laser_ops_ef[4] = {
+		SC_AC_OP_READ, SC_AC_OP_WRITE, SC_AC_OP_ADMIN, SC_AC_OP_DELETE_SELF};
+unsigned char laser_ops_do[3] = {
+		SC_AC_OP_READ, SC_AC_OP_WRITE, SC_AC_OP_ADMIN};
+unsigned char laser_ops_ko[7] = {
+		SC_AC_OP_READ, SC_AC_OP_UPDATE, SC_AC_OP_ADMIN, SC_AC_OP_DELETE_SELF, SC_AC_OP_GENERATE, SC_AC_OP_PIN_RESET, SC_AC_OP_CRYPTO};
+unsigned char laser_ops_pin[7] = {
+		SC_AC_OP_READ, SC_AC_OP_PIN_CHANGE, SC_AC_OP_ADMIN, SC_AC_OP_DELETE_SELF, SC_AC_OP_GENERATE, SC_AC_OP_PIN_RESET, SC_AC_OP_CRYPTO};
+
+static const unsigned char laser_sha1_digest_pref[] = {0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2B, 0x0E, 0x03, 0x02, 0x1A, 0x05, 0x00, 0x04, 0x14};
+static const unsigned char laser_sha224_digest_pref[] = {0x30, 0x2D, 0x30, 0x0D, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x04, 0x05, 0x00, 0x04, 0x1C};
+static const unsigned char laser_sha256_digest_pref[] = {0x30, 0x31, 0x30, 0x0D, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20};
+static const unsigned char laser_sha384_digest_pref[] = {0x30, 0x41, 0x30, 0x0D, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0x04, 0x30};
+static const unsigned char laser_sha512_digest_pref[] = {0x30, 0x51, 0x30, 0x0D, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0x04, 0x40};
```

Comment:
These are defined in libopensc/padding as they are standard digestinfo prefixes. Please, reuse them from there if possible to avoid code duplication.